planner/core/rule_partition_processor.go

```diff
@@ -624,3 +651,139 @@ func (s *partitionProcessor) makeUnionAllChildren(ds *DataSource, pi *model.Part
+	}
+	return s.makeUnionAllChildren(ds, pi, result)
+}
+
+var _ partitionRangePrunner = &partitionRangePrunner2{}
+
+// partitionRangePrunner2 is used by 'partition by range columns'.
+type partitionRangePrunner2 struct {
+	data     []expression.Expression
+	partCol  *expression.Column
+	maxvalue bool
+}
+
+func makePartitionRangePrunner2(ds *DataSource, pi *model.PartitionInfo) (*partitionRangePrunner2, error) {
+	var maxValue bool
+	data := make([]expression.Expression, len(pi.Definitions))
+	schema := expression.NewSchema(ds.TblCols...)
+	exprs, err := expression.ParseSimpleExprsWithNames(ds.ctx, pi.Columns[0].L, schema, ds.names)
+	if err != nil {
+		return nil, err
+	}
+	partCol := exprs[0].(*expression.Column)
+	for i := 0; i < len(pi.Definitions); i++ {
+		if strings.EqualFold(pi.Definitions[i].LessThan[0], "MAXVALUE") {
+			// Use a bool flag instead of math.MaxInt64 to avoid the corner cases.
+			maxValue = true
+		} else {
+			tmp, err := expression.ParseSimpleExprsWithNames(ds.ctx, pi.Definitions[i].LessThan[0], schema, ds.names)
+			if err != nil {
+				return nil, err
+			}
+			data[i] = tmp[0]
+		}
+	}
+	return &partitionRangePrunner2{data, partCol, maxValue}, nil
+}
+
+func (p *partitionRangePrunner2) fullRange() partitionRangeOR {
+	return fullRange(len(p.data))
+}
+
+func (p *partitionRangePrunner2) partitionRangeForExpr(sctx sessionctx.Context, expr expression.Expression) (int, int, bool) {
+	op, ok := expr.(*expression.ScalarFunction)
+	if !ok {
+		return 0, len(p.data), false
+	}
+
+	switch op.FuncName.L {
+	case ast.EQ, ast.LT, ast.GT, ast.LE, ast.GE:
+	case ast.IsNull:
+		// isnull(col)
+		if arg0, ok := op.GetArgs()[0].(*expression.Column); ok && arg0.ID == p.partCol.ID {
+			return 0, 1, true
+		}
+		return 0, len(p.data), false
+	default:
+		return 0, len(p.data), false
+	}
+	opName := op.FuncName.L
+
+	var col *expression.Column
+	var con *expression.Constant
+	if arg0, ok := op.GetArgs()[0].(*expression.Column); ok && arg0.ID == p.partCol.ID {
+		if arg1, ok := op.GetArgs()[1].(*expression.Constant); ok {
+			col, con = arg0, arg1
+		}
+	} else if arg0, ok := op.GetArgs()[1].(*expression.Column); ok && arg0.ID == p.partCol.ID {
+		if arg1, ok := op.GetArgs()[0].(*expression.Constant); ok {
+			opName = opposite(opName)
+			col, con = arg0, arg1
+		}
+	}
+	if col == nil || con == nil {
+		return 0, len(p.data), false
+	}
+
+	start, end := p.pruneUseBinarySearch(sctx, opName, con)
+	return start, end, true
+}
+
+func (p *partitionRangePrunner2) pruneUseBinarySearch(sctx sessionctx.Context, op string, data *expression.Constant) (start int, end int) {
+	var err error
+	var isNull bool
+	compare := func(ith int, op string, v *expression.Constant) bool {
+		if ith == len(p.data)-1 {
+			if p.maxvalue {
+				return true
+			}
+		}
+		var expr expression.Expression
+		expr, err = expression.NewFunction(sctx, op, types.NewFieldType(mysql.TypeLonglong), p.data[ith], v)
```

Comment:
>both RANGE COLUMNS partitioning and LIST COLUMNS partitioning support the use of non-integer columns for defining value ranges or list members. The permitted data types are shown in the following list:
>
> - All integer types: TINYINT, SMALLINT, MEDIUMINT, INT (INTEGER), and BIGINT. (This is the same as with partitioning by RANGE and LIST.)
>
> - Other numeric data types (such as DECIMAL or FLOAT) are not supported as partitioning columns.
>
> - DATE and DATETIME.
>
> - Columns using other data types relating to dates or times are not supported as partitioning columns.
>
> The following string types: CHAR, VARCHAR, BINARY, and VARBINARY.
>
>TEXT and BLOB columns are not supported as partitioning columns.

it seems `partition by range columns` only support three type int/date/string, so ideal way is impl two addition lessThanData type ---------- `lessThanDataInt`, `lessThanDataTime` and `lessThanDataString` to void new & eval expression for each compare?

but LGTM to use current impl to quick fix

Reply:
If some type is not supported, we should prevent the case in DDL check, rather than in partition pruning.
`lessThanDataTime` and `lessThanDataString` seems good, we can optimize it later.